modules/dnn/src/layers/elementwise_layers.cpp

```diff
@@ -654,11 +654,19 @@ struct SwishFunctor : public BaseFunctor
 struct MishFunctor : public BaseFunctor
 {
     typedef MishLayer Layer;
+    float MISH_THRESHOLD;
```

Comment:
What is the source of this new parameter?

Reply:
https://github.com/AlexeyAB/darknet/blob/36c73c5b9e3f2e72049fb68566e32632f6c70e85/src/activations.c#L150

Reply:
Do we still need threshold or can revert layer?

Reply:
`MISH_THRESHOLD` is simply a cutoff for switching from a numerically unstable expression to a numerically stable expression in softplus.

The current OpenCV implementation which is `x * tanh(log(1 + exp(x)))` is numerically unstable. This implementation goes to zero when x is more negative than `-16`. Using `std::log1p` instead of `std::log(1 + std::exp(x))` might improve the numerical accuracy but this will still suffer as numbers become even more negative.

Darknet has been using `log(1 + exp(x))` and it seems to be very inaccurate (and so will OCV CPU implementation) for numbers smaller than `-2`. I have dumped the values [here](https://gist.github.com/YashasSamaga/3fdf001d32f04062e3f36495d5c962db) comparing darknet with [other implementations](https://cs.stackexchange.com/questions/125002/fast-and-stable-x-tanhlog1pexpx-computation).

If I remember correctly, the dynamic range of activations in YOLOv4 seems to cover the range ~`[-12000, 2000]` for a bunch of random 10 natural images. The middle 50% is covered in ~`[-100, 100]`.

Both Darknet and OpenCV CPU are giving really good results with their implementations. I don't know if there is a need to have a numerically stable implementation. But just for record sake, I have made a reply here.

Reply:
Now we use such Mish-implementation  https://github.com/AlexeyAB/darknet/issues/5452#issuecomment-627414024
```cpp
 __device__ float mish_yashas(float x) 
 { 
     float e = __expf(x); 
     if (x <= -18.0f) 
         return x * e; 
  
     float n = e * e + 2 * e; 
     if (x <= -5.0f) 
         return x * __fdividef(n, n + 2); 
  
     return x - 2 * __fdividef(x, n + 2); 
 } 
```
And get:
* +3% speedup
* the same accuracy on MSCOCO (testdev) yolov4.cfg 512x512 - 43.0% AP, 64.9% AP50
```
overall performance
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.430
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.649
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.465
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.243
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.461
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.552
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.341
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.542
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.572
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.375
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.612
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.729
Done (t=506.74s)
```